File: backend/app/models/stock_movement.py
import uuid
from datetime import datetime
from sqlalchemy import Column, String, Numeric, DateTime, ForeignKey, Text, Integer
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from ..database import Base


class StockMovement(Base):
    __tablename__ = "stock_movements"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    product_id = Column(UUID(as_uuid=True), ForeignKey("products.id"), nullable=False)
    movement_type = Column(String(20), nullable=False)
    quantity = Column(Numeric(12, 3), nullable=False)
    stock_before = Column(Numeric(12, 3), nullable=False)
    stock_after = Column(Numeric(12, 3), nullable=False)
    reference_id = Column(UUID(as_uuid=True), nullable=True)
    reference_type = Column(String(50), nullable=True)
    notes = Column(Text, nullable=True)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow)

    product = relationship("Product", back_populates="stock_movements")
    user = relationship("User")

    def to_dict(self):
        return {
            "id": str(self.id),
            "product_id": str(self.product_id),
            "movement_type": self.movement_type,
            "quantity": float(self.quantity),
            "stock_before": float(self.stock_before),
            "stock_after": float(self.stock_after),
            "reference_id": str(self.reference_id) if self.reference_id else None,
            "reference_type": self.reference_type,
            "notes": self.notes,
            "user_id": str(self.user_id),
            "created_at": self.created_at.isoformat() if self.created_at else None,
        }

File: fruit_pos/lib/models/stock_movement.dart
class StockMovement {
  final String id;
  final String productId;
  final String? productName;
  final String movementType;
  final double quantity;
  final double stockBefore;
  final double stockAfter;
  final String? referenceId;
  final String? referenceType;
  final String? notes;
  final String userId;
  final String? createdAt;

  StockMovement({
    required this.id,
    required this.productId,
    this.productName,
    required this.movementType,
    required this.quantity,
    required this.stockBefore,
    required this.stockAfter,
    this.referenceId,
    this.referenceType,
    this.notes,
    required this.userId,
    this.createdAt,
  });

  factory StockMovement.fromJson(Map<String, dynamic> json) => StockMovement(
        id: json['id']?.toString() ?? '',
        productId: json['product_id']?.toString() ?? '',
        productName: json['product_name']?.toString(),
        movementType: json['movement_type']?.toString() ?? '',
        quantity: (json['quantity'] ?? 0).toDouble(),
        stockBefore: (json['stock_before'] ?? 0).toDouble(),
        stockAfter: (json['stock_after'] ?? 0).toDouble(),
        referenceId: json['reference_id']?.toString(),
        referenceType: json['reference_type']?.toString(),
        notes: json['notes']?.toString(),
        userId: json['user_id']?.toString() ?? '',
        createdAt: json['created_at']?.toString(),
      );
}